code/datums/components/food/golem_food.dm

```diff
@@ -0,0 +1,112 @@
+/// Component which allows something to be eaten by a golem
+/datum/component/golem_food
+	/// Whether to destroy our item when we eat it
+	var/consume_on_eat
+	/// Reference to the kind of food we provide
+	var/datum/golem_food_buff/snack_type
+	/// Our abstract fake food item
+	var/obj/item/food/golem_food/golem_snack
+	/// Any extra checks which need to be done when seeing if this is edible
+	var/datum/callback/extra_validation
+
+/datum/component/golem_food/Initialize(consume_on_eat = TRUE, golem_food_key, datum/callback/extra_validation)
+	if (!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	if (!golem_food_key || !is_path_in_list(golem_food_key, GLOB.golem_stack_food_directory))
+		CRASH("Golem food type not specified, this is a required parameter.")
+
+	src.consume_on_eat = consume_on_eat
+	snack_type = GLOB.golem_stack_food_directory[golem_food_key]
+	src.extra_validation = extra_validation
+
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK, PROC_REF(on_attack))
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, PROC_REF(on_examine))
+
+/datum/component/golem_food/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_ITEM_ATTACK, COMSIG_PARENT_EXAMINE))
+	return ..()
+
+/datum/component/golem_food/Destroy(force, silent)
+	QDEL_NULL(golem_snack)
+	QDEL_NULL(extra_validation)
+	return ..()
+
+/// Attempt to feed this item to golem
+/datum/component/golem_food/proc/on_attack(atom/source, mob/living/target, mob/living/user, click_parameters)
+	SIGNAL_HANDLER
+	if (user.combat_mode || !HAS_TRAIT(target, TRAIT_ROCK_EATER))
+		return
+	if (extra_validation && !extra_validation.Invoke())
+		source.balloon_alert(user, "not edible!")
+		return COMPONENT_CANCEL_ATTACK_CHAIN
+	if (!snack_type.can_consume(target))
+		source.balloon_alert(user, "incompatible mineral!")
+		return COMPONENT_CANCEL_ATTACK_CHAIN
+	if (!golem_snack)
+		golem_snack = new(
+			/* loc = */ null,
+			/* name = */ source.name,
+			/* consume_food = */ consume_on_eat,
+			/* food_buff = */ snack_type,
+			/* owner = */ parent,
```

Comment:
why have the args formatted like this, do named args not work on new()? if they don't this is fine but i've never seen this patterh before lol

Reply:
Afraid so.

Reply:
I think this is mentioned somewhere in our standards as an alternative for this case, or at least I remember it being written somewhere...